File: app/src/main/java/xyz/rodit/snapmod/features/opera/MenuModifier.java
package xyz.rodit.snapmod.features.opera;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import xyz.rodit.snapmod.features.Contextual;
import xyz.rodit.snapmod.features.FeatureContext;

public class MenuModifier extends Contextual implements OperaPlugin {

    private static final String KEY_NAME = "action_menu_options";

    private final List<MenuPlugin> plugins = new ArrayList<>();

    public MenuModifier(FeatureContext context) {
        super(context);

        plugins.add(new SaveMenuOption(context));
    }

    @Override
    public boolean isEnabled() {
        return true;
    }

    @Override
    public boolean shouldOverride(String key) {
        return KEY_NAME.equals(key);
    }

    @Override
    public Object override(String key, Object value) {
        List newList = new ArrayList();
        newList.addAll(((List) value));
        newList.addAll(plugins.stream().filter(MenuPlugin::isEnabled).flatMap(p -> p.createActions().stream().map(a -> a.instance)).collect(Collectors.toList()));
        return newList;
    }
}

File: app/src/main/java/xyz/rodit/snapmod/features/opera/MenuModifier.kt
package xyz.rodit.snapmod.features.opera

import xyz.rodit.snapmod.features.Contextual
import xyz.rodit.snapmod.features.FeatureContext
import xyz.rodit.snapmod.mappings.OperaActionMenuOptionViewModel
import xyz.rodit.snapmod.mappings.ParamsMap

private const val KEY_NAME = "action_menu_options"

class MenuModifier(context: FeatureContext) : Contextual(context), OperaPlugin {

    private val plugins: MutableList<MenuPlugin> = ArrayList()

    override val isEnabled: Boolean
        get() {
            return true
        }

    override fun shouldOverride(params: ParamsMap, key: String): Boolean {
        return KEY_NAME == key
    }

    override fun override(params: ParamsMap, key: String, value: Any): Any {
        if (value !is List<*>) return value
        val newList: MutableList<Any> = ArrayList()
        newList.addAll(value.filterNotNull())
        newList.addAll(
            plugins.filter(MenuPlugin::isEnabled)
                .flatMap { it.createActions(params) }
                .map(OperaActionMenuOptionViewModel::instance)
        )
        return newList
    }

    init {
        plugins.add(SaveMenuOption(context))
        plugins.add(context.features.get<CustomStoryOptions>())
    }
}